File: apps/core/src/main/java/bada/quantum/Quintic.java
package bada.quantum;

/**
 * The quintic (5次方程式) carrier. Abel–Ruffini denies a radical solution
 * formula; we solve numerically with Durand–Kerner and use the fifth roots of
 * unity (period 2π/5) as a quinary carrier whose phase is non-linearly warped
 * by the pair correlation — period-tuned yet non-linear.
 */
public final class Quintic {
    public static final double TWO_PI = 2.0 * Math.PI;
    public static final double PERIOD = TWO_PI / 5.0;

    /** The five fifth-roots of unity exp(2*pi*i*k/5). */
    public static Complex[] rootsOfUnity() {
        Complex[] r = new Complex[5];
        for (int k = 0; k < 5; k++) {
            r[k] = new Complex(Math.cos(PERIOD * k), Math.sin(PERIOD * k));
        }
        return r;
    }

    /** Evaluate x^5 + c4 x^4 + c3 x^3 + c2 x^2 + c1 x + c0. */
    static Complex poly5(Complex x, Complex[] c) {
        return x.pow(5)
                .add(c[0].mul(x.pow(4)))
                .add(c[1].mul(x.pow(3)))
                .add(c[2].mul(x.pow(2)))
                .add(c[3].mul(x))
                .add(c[4]);
    }

    /** Durand–Kerner solver for a monic quintic given [c4,c3,c2,c1,c0]. */
    public static Complex[] solve(double[] coeffs, int iterations) {
        Complex[] c = new Complex[5];
        for (int i = 0; i < 5; i++) c[i] = Complex.real(coeffs[i]);
        Complex seed = new Complex(0.4, 0.9);
        Complex[] roots = new Complex[5];
        for (int k = 0; k < 5; k++) roots[k] = seed.pow(k);
        for (int it = 0; it < iterations; it++) {
            Complex[] next = roots.clone();
            for (int i = 0; i < 5; i++) {
                Complex num = poly5(roots[i], c);
                Complex den = new Complex(1, 0);
                for (int j = 0; j < 5; j++) if (j != i) den = den.mul(roots[i].sub(roots[j]));
                next[i] = roots[i].sub(num.div(den));
            }
            roots = next;
        }
        return roots;
    }

    public static Complex[] solve(double[] coeffs) { return solve(coeffs, 200); }

    /** Non-linear carrier phase for the k-th quinary symbol at correlation E. */
    public static double carrierPhase(int k, double correlation) {
        double base = PERIOD * (((k % 5) + 5) % 5);
        return base + correlation * Math.sin(base);
    }

    /** The quinary constellation for a given pair correlation. */
    public static Complex[] constellation(double correlation) {
        Complex[] c = new Complex[5];
        for (int k = 0; k < 5; k++) {
            double ph = carrierPhase(k, correlation);
            c[k] = new Complex(Math.cos(ph), Math.sin(ph));
        }
        return c;
    }
}

File: bada_ruby/lib/bada/quantum/quintic.rb
# frozen_string_literal: true

module Bada
  module Quantum
    # The quintic (5次方程式) carrier.
    #
    # Abel–Ruffini says the general quintic has no *radical* solution formula;
    # its roots are given instead by a genuinely non-linear "solution formula"
    # (Hermite's elliptic / Bring-radical form). We do not need the elliptic
    # machinery here — we need its two structural gifts:
    #
    #   1. a **period**: the reduced quintic  z^5 = 1  has its five roots equally
    #      spaced by  2*pi/5  on the unit circle (the fifth roots of unity). That
    #      spacing is the natural period of a 5-symbol (quinary) telegraph.
    #   2. a **non-linear modulation**: because the map is not solvable by
    #      radicals, the phase advance per symbol is a non-linear function of the
    #      pair's entanglement correlation, not a plain linear ramp.
    #
    # We solve arbitrary quintics numerically (Durand–Kerner / Weierstrass) to
    # keep the "solution formula" honest and testable, and use the roots as the
    # carrier constellation for the space telegraph.
    module Quintic
      module_function

      TWO_PI = 2.0 * Math::PI
      PERIOD = TWO_PI / 5.0 # spacing of the fifth roots of unity

      # The five fifth-roots of unity  exp(2*pi*i*k/5), k = 0..4.
      def roots_of_unity
        (0...5).map { |k| Complex(Math.cos(PERIOD * k), Math.sin(PERIOD * k)) }
      end

      # Durand–Kerner solver for a monic quintic given its lower coefficients
      # [c4, c3, c2, c1, c0] of  x^5 + c4 x^4 + ... + c0.  Returns 5 complex roots.
      def solve(coeffs, iterations: 200)
        c = coeffs.map { |v| Complex(v, 0) }
        # initial guesses on a circle (classic Durand–Kerner seed 0.4+0.9i)
        seed = Complex(0.4, 0.9)
        roots = (0...5).map { |k| seed**k }
        iterations.times do
          new_roots = roots.dup
          5.times do |i|
            num = poly5(roots[i], c)
            den = Complex(1, 0)
            5.times { |j| den *= (roots[i] - roots[j]) if j != i }
            new_roots[i] = roots[i] - num / den
          end
          roots = new_roots
        end
        roots
      end

      # Evaluate  x^5 + c4 x^4 + c3 x^3 + c2 x^2 + c1 x + c0.
      def poly5(x, c)
        x**5 + c[0] * x**4 + c[1] * x**3 + c[2] * x**2 + c[3] * x + c[4]
      end

      # Bring–Jerrard style principal quintic  x^5 + p*x + q  (the form Hermite
      # solved with elliptic functions). Its discriminant governs how many real
      # roots exist, i.e. the "period" of sign changes along the real axis.
      def bring_jerrard(p, q)
        solve([0.0, 0.0, 0.0, p, q])
      end

      # Non-linear phase for the k-th quinary symbol given the entanglement
      # correlation E in [-1,1]. The base advance is the periodic k*2*pi/5; the
      # correlation warps it non-linearly (sin of the base phase), so the carrier
      # is *tuned to the equation's period* yet non-linear — exactly the
      # "方程式の周期に合わせた非線形の性質".
      def carrier_phase(k, correlation)
        base = PERIOD * (k % 5)
        base + correlation * Math.sin(base)
      end

      # The full quinary carrier constellation for a given pair correlation.
      def constellation(correlation)
        (0...5).map do |k|
          ph = carrier_phase(k, correlation)
          Complex(Math.cos(ph), Math.sin(ph))
        end
      end
    end
  end
end
